From 5128f1f9990c1714ecccaf09419106407ea96c2f Mon Sep 17 00:00:00 2001
From: Kuppuswamy Sathyanarayanan <sathyanarayanan.kuppuswamy@intel.com>
Date: Wed, 23 Jul 2014 22:44:53 -0700
Subject: [PATCH 2/2] dw_i2c: Handle errors when STOP condition has occured on
 I2C interface

rx_stop_det status bit  will be set in interrupt status register
when a  STOP condition has occurred on the I2C interface regardless
of whether I2C is operating  in slave or master mode. This patch
cleans the RX FIFO during this scenario.

Signed-off-by: Kuppuswamy Sathyanarayanan <sathyanarayanan.kuppuswamy@intel.com>
---
 drivers/i2c/busses/i2c-designware-core.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/busses/i2c-designware-core.c b/drivers/i2c/busses/i2c-designware-core.c
index 05496e45e71d..4bd73abbe71b 100644
--- a/drivers/i2c/busses/i2c-designware-core.c
+++ b/drivers/i2c/busses/i2c-designware-core.c
@@ -818,8 +818,17 @@ irqreturn_t i2c_dw_isr(int this_irq, void *dev_id)
 	 */
 
 tx_aborted:
-	if ((stat & (DW_IC_INTR_TX_ABRT | DW_IC_INTR_STOP_DET)) || dev->msg_err)
+	if ((stat & (DW_IC_INTR_TX_ABRT | DW_IC_INTR_STOP_DET)) || dev->msg_err) {
+		/*
+		 * Check DW_IC_RXFLR register and
+		 * read from the RX FIFO if it's not
+		 * empty.
+		 */
+		if ((stat & DW_IC_INTR_STOP_DET) &&
+			dw_readl(dev, DW_IC_RXFLR) > 0)
+			i2c_dw_read(dev);
 		complete(&dev->cmd_complete);
+	}
 
 	return IRQ_HANDLED;
 }
-- 
1.7.9.5

